File: backends/dotnet/shared/Implementations/Messaging/Messaging.RabbitMQ/Handlers/Q/Ping.cs
// -----------------------------------------------------------------------
// <copyright file="Ping.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Messaging.RabbitMQ.Handlers.Q;

using System.Diagnostics;
using D2.Shared.Handler;
using D2.Shared.Result;
using global::RabbitMQ.Client;
using Microsoft.Extensions.Logging;

using H = D2.Shared.Interfaces.Messaging.Handlers.Q.IRead.IPingHandler;
using I = D2.Shared.Interfaces.Messaging.Handlers.Q.IRead.PingInput;
using O = D2.Shared.Interfaces.Messaging.Handlers.Q.IRead.PingOutput;

/// <summary>
/// Handler for pinging RabbitMQ to verify connectivity.
/// </summary>
public partial class Ping : BaseHandler<H, I, O>, H
{
    private readonly IConnection r_connection;

    /// <summary>
    /// Initializes a new instance of the <see cref="Ping"/> class.
    /// </summary>
    ///
    /// <param name="connection">
    /// The RabbitMQ connection.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Ping(
        IConnection connection,
        IHandlerContext context)
        : base(context)
    {
        r_connection = connection;
    }

    /// <inheritdoc/>
    protected override ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            var isOpen = r_connection.IsOpen;
            sw.Stop();

            if (isOpen)
            {
                return ValueTask.FromResult(
                    D2Result<O?>.Ok(new O(true, sw.ElapsedMilliseconds, null)));
            }

            LogConnectionNotOpen(Context.Logger, TraceId);

            return ValueTask.FromResult(
                D2Result<O?>.Ok(new O(false, sw.ElapsedMilliseconds, "Connection is not open")));
        }
        catch (Exception ex)
        {
            sw.Stop();
            LogPingFailed(Context.Logger, ex, TraceId);

            return ValueTask.FromResult(
                D2Result<O?>.Ok(new O(false, sw.ElapsedMilliseconds, ex.Message)));
        }
    }

    /// <summary>
    /// Logs that the RabbitMQ connection is not open.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Warning, Message = "RabbitMQ connection is not open. TraceId: {TraceId}")]
    private static partial void LogConnectionNotOpen(ILogger logger, string? traceId);

    /// <summary>
    /// Logs that the RabbitMQ ping failed with an exception.
    /// </summary>
    [LoggerMessage(EventId = 2, Level = LogLevel.Error, Message = "RabbitMQ ping failed. TraceId: {TraceId}")]
    private static partial void LogPingFailed(ILogger logger, Exception ex, string? traceId);
}

File: backends/node/shared/messaging/src/handlers/q/ping.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { Messaging } from "@d2/interfaces";
import type { MessageBus } from "../../message-bus.js";

type Input = Messaging.PingInput;
type Output = Messaging.PingOutput;

export class PingMessageBus extends BaseHandler<Input, Output> implements Messaging.IPingHandler {
  private readonly messageBus: MessageBus;

  constructor(messageBus: MessageBus, context: IHandlerContext) {
    super(context);
    this.messageBus = messageBus;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    const start = performance.now();
    try {
      const connected = await this.messageBus.ping();
      const latencyMs = Math.round(performance.now() - start);
      if (connected) {
        return D2Result.ok({ data: { healthy: true, latencyMs } });
      }
      return D2Result.ok({
        data: { healthy: false, latencyMs, error: "Connection not ready" },
      });
    } catch (err) {
      const latencyMs = Math.round(performance.now() - start);
      const error = err instanceof Error ? err.message : "Unknown error";
      return D2Result.ok({ data: { healthy: false, latencyMs, error } });
    }
  }
}
